handlers/update.go

```diff
@@ -28,81 +28,123 @@ func MakeUpdateHandler(functionNamespace string, clientset *kubernetes.Clientset
-					labels[k] = v
-				}
-			}
+		deployment.Spec.Template.Spec.NodeSelector = createSelector(request.Constraints)
 
-			deployment.Labels = labels
-			deployment.Spec.Template.ObjectMeta.Labels = labels
+		labels := map[string]string{
+			"faas_function": request.Service,
+			"uid":           fmt.Sprintf("%d", time.Now().Nanosecond()),
+		}
 
-			resources, resourceErr := createResources(request)
-			if resourceErr != nil {
-				w.WriteHeader(http.StatusBadRequest)
-				w.Write([]byte(resourceErr.Error()))
-				return
+		if request.Labels != nil {
+			if min := getMinReplicaCount(*request.Labels); min != nil {
+				deployment.Spec.Replicas = min
 			}
 
-			deployment.Spec.Template.Spec.Containers[0].Resources = *resources
-
-			existingSecrets, err := getSecrets(clientset, functionNamespace, request.Secrets)
-			if err != nil {
-				w.WriteHeader(http.StatusBadRequest)
-				w.Write([]byte(err.Error()))
-				return
+			for k, v := range *request.Labels {
+				labels[k] = v
 			}
+		}
 
-			err = UpdateSecrets(request, deployment, existingSecrets)
-			if err != nil {
-				log.Println(err)
-				w.WriteHeader(http.StatusBadRequest)
-				w.Write([]byte(err.Error()))
-				return
-			}
+		deployment.Labels = labels
+		deployment.Spec.Template.ObjectMeta.Labels = labels
+
+		deployment.Annotations = annotations
+		deployment.Spec.Template.Annotations = annotations
+		deployment.Spec.Template.ObjectMeta.Annotations = annotations
+
+		resources, resourceErr := createResources(request)
+		if resourceErr != nil {
+			return resourceErr, http.StatusBadRequest
 		}
 
-		if _, updateErr := clientset.ExtensionsV1beta1().
-			Deployments(functionNamespace).
-			Update(deployment); updateErr != nil {
+		deployment.Spec.Template.Spec.Containers[0].Resources = *resources
 
-			w.WriteHeader(http.StatusInternalServerError)
-			w.Write([]byte(updateErr.Error()))
+		existingSecrets, err := getSecrets(clientset, functionNamespace, request.Secrets)
+		if err != nil {
+			return err, http.StatusBadRequest
 		}
+
+		err = UpdateSecrets(request, deployment, existingSecrets)
+		if err != nil {
+			log.Println(err)
+			return err, http.StatusBadRequest
+		}
+	}
+
+	if _, updateErr := clientset.ExtensionsV1beta1().
+		Deployments(functionNamespace).
+		Update(deployment); updateErr != nil {
+
+		return updateErr, http.StatusInternalServerError
+	}
+
+	return nil, http.StatusOK
+}
+
+func updateService(
+	functionNamespace string,
+	clientset *kubernetes.Clientset,
+	request requests.CreateFunctionRequest,
+	annotations map[string]string) (err error, httpStatus int) {
+
+	getOpts := metav1.GetOptions{}
+
+	service, findServiceErr := clientset.CoreV1().
+		Services(functionNamespace).
+		Get(request.Service, getOpts)
+
+	if findServiceErr != nil {
+		return findServiceErr, http.StatusNotFound
+	}
+
+	service.Annotations = annotations
+
+	if _, updateErr := clientset.CoreV1().
+		Services(functionNamespace).
+		Update(service); updateErr != nil {
+
```

Comment:
@LucasRoesler @alexellis i've *not* added a test to this method as its not unit testable, it calls the k8s api